change-validation/src/main/scala/com/normation/plugins/changevalidation/ValidationNeeded.scala

```diff
@@ -0,0 +1,137 @@
+ * arbitrary rules defined in implementation.
+ */
+trait ValidationNeeded {
+  def forRule       (actor: EventActor, change: RuleChangeRequest       ): Box[Boolean]
+  def forDirective  (actor: EventActor, change: DirectiveChangeRequest  ): Box[Boolean]
+  def forNodeGroup  (actor: EventActor, change: NodeGroupChangeRequest  ): Box[Boolean]
+  def forGlobalParam(actor: EventActor, change: GlobalParamChangeRequest): Box[Boolean]
+}
+
+
+/*
+ * A version of the "validationNeeded" plugin which bases its oracle on a list
+ * of group. The list of group is used to mark nodes.
+ *
+ * Any modification that impacts a marked node triggers a validation:
+ * - a modification in a GlobalParam is always validated (because can be used in CFEngine code)
+ * - a modification in a node group is validated only if it's one of the supervised one
+ * - a modification in a rule is validated only if:
+ *    - one of the modification add or remove a monitored group
+ *    - intersection of the set of nodes targeted by the rule with the set of marked nodes is non empty
+ * - a modification in a directive is validated if it as at least configured in one rule where modification
+ *   are supervised.
+ *
+ */
+class NodeGroupValidationNeeded(
+    monitoredTargets: () => Box[Set[SimpleTarget]]
+  , repos           : RoChangeRequestRepository
+  , ruleLib         : RoRuleRepository
+  , groupLib        : RoNodeGroupRepository
+  , nodeInfoService : NodeInfoService
+) extends ValidationNeeded {
+
+  /*
+   * A rule need external validation if any of its current or futur nodes
+   * are also in a monitored group.
+   * We can't check only for group target by the rule, because of (for ex)
+   * that case:
+   * - A is a node in Group1 (supervised) and Group2 (not supervised)
+   * - rule R is changed to add Group2 in its target (or oposite change: Group2 removed)
+   * - the change must be validated.
+   */
+  override def forRule(actor: EventActor, change: RuleChangeRequest): Box[Boolean] = {
+    val start = System.currentTimeMillis()
+    val res = for {
+      groups    <- groupLib.getFullGroupLibrary()
+      nodeInfo  <- nodeInfoService.getAll()
+      monitored <- monitoredTargets()
+    } yield {
+      val targets = Set(change.newRule) ++ change.previousRule.toSet
+      checkNodeTargetByRule(groups, nodeInfo, monitored, targets)
+    }
+    ChangeValidationLogger.Metrics.debug(s"Check rule '${change.newRule.name}' [${change.newRule.id.value}] change requestion need for validation in ${System.currentTimeMillis() - start}ms")
+    res
+  }
+
+  /**
+   * This method checks if at least one of the nodes belonging to rule targets for the change
+   * is supervised.
+   */
+  def checkNodeTargetByRule(groups: FullNodeGroupCategory, allNodeInfo: Map[NodeId, NodeInfo], monitored: Set[SimpleTarget], rules: Set[Rule]): Boolean = {
+    val monitoredNodes = groups.getNodeIds(monitored.map(identity), allNodeInfo)
```

Comment:
Why do you need to '.map(identity)' ?

Reply:
because set is monovariant, and I have a Set[SimpleTarget] where I need a Set[RuleTarget].